README.md
Description
===========

[![Build Status](https://travis-ci.org/aioi/chef-freetds.png?branch=master)](https://travis-ci.org/aioi/chef-freetds)

This cookbook install and configure FreeTDS

Platform
--------

* CentOS (tested on 6.6 - source only)
* Debian (tested on 7.4)
* Ubuntu (tested on 12.04 and 14.04)

May work with or without modification on other distributions.

Usage
-----
Just include `freetds` in your node's `run_list`.

If you want to configure some server add them to your node's attributes:

```ruby
    node['freetds']['servers'] = [
      {
        'name' => 'egServer70',
        'description' => 'A typical Microsoft server',
        'host'=> 'ntmachine.domain.com',
        'port' => 1433,
        'tds_version' => '7.0',
        'text_size' => 1024,
        'client_charset' => 'UTF-8',
        'encryption' => 'off' # require/request/off
      }
    ]
```

Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: [Olivier Brisse](https://github.com/ouranos)


CHANGELOG.md
# CHANGELOG for freetds

This file is used to list changes made in each version of freetds.

## master (unreleased)

## 0.3.0 (2016-08-30):

* Add support for client_charset definition [\#3](https://github.com/aioi/chef-freetds/pull/3) ([jakimowicz](https://github.com/jakimowicz))
* Add support for encryption [\#6](https://github.com/aioi/chef-freetds/pull/6) ([amf9t2](https://github.com/amf9t2))
* Default to 0.95.87 when compiling from source [\#5](https://github.com/aioi/chef-freetds/pull/5) ([l8nite](https://github.com/l8nite))
* Update foodcritic (0.4.0) and rubocop (0.37.2)
* Add Travis CI build on ruby 2.0, 2.2 & 2.3

## 0.2.0 (2012-12-13):

* Add a package recipe and switch default install mode to package
* Fix template (use text_size, tds_version)
* Add CentOS tests

## 0.1.4:

* Add integration (test-kitchen) and style tests (rubocop, foodcritic)
* Travis CI integration

## 0.1.3:

* Fix cookbook name in metadata

## 0.1.1:

* Fix licence

## 0.1.0:

* Initial release of freetds


test/integration/source/serverspec/default_spec.rb
require 'serverspec'

set :backend, :exec
set :path, '/usr/local/bin:$PATH'

describe 'freetds::default' do
  describe file('/usr/local/etc/freetds.conf') do
    it { should be_file }
  end

  describe command('which tsql') do
    its(:stdout) { should match '/usr/local/bin/tsql' }
    its(:exit_status) { should eq 0 }
  end

  describe command('tsql -C') do
    its(:stdout) { should match 'Version: freetds v0.91' }
    its(:stdout) { should match 'TDS version: 7.1' }
    its(:exit_status) { should eq 0 }
  end
end


test/integration/default/serverspec/default_spec.rb
require 'serverspec'

set :backend, :exec

describe 'freetds::default' do
  describe file('/usr/local/etc/freetds.conf') do
    it { should be_file }
  end

  describe command('which tsql') do
    its(:stdout) { should match '/usr/local/bin/tsql' }
    its(:exit_status) { should eq 0 }
  end

  describe command('tsql -C') do
    its(:stdout) { should match 'Version: freetds v0.91' }
    its(:stdout) { should match 'TDS version: 7.1' }
    its(:exit_status) { should eq 0 }
  end
end


metadata.rb
name             'freetds'
maintainer       'Olivier Brisse'
maintainer_email 'obrisse@aioinissaydowa.com.au'
license          'Apache 2.0'
description      'Installs/Configures FreeTDS'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.3.0'

%w(ubuntu debian centos).each do |os|
  supports os
end

depends 'apt'
depends 'build-essential'

attribute 'version',
  :display_name => 'Version',
  :description => 'The version of FreeTDS to install',
  :default => '0.91'

attribute 'checksum',
  :display_name => 'Checksum',
  :description => 'checksum of the FreeTDS archive',
  :required => 'optional'

attribute 'tds_version',
  :display_name => 'TDS Version',
  :description => 'TDS protocol version',
  :default => '7.1'

attribute 'odbc',
  :display_name => 'ODBC support',
  :description => 'Select wether or not to build the ODBC driver',
  :default => 'false'

attribute 'servers',
  :display_name => 'Server list',
  :description => 'See README'


recipes/source.rb
#
# Cookbook Name:: freetds
# Recipe:: source
#
# Copyright (C) 2014 Olivier Brisse
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'build-essential::default'

version = node['freetds']['version']
freetds_url = node['freetds']['url'] ||
              "ftp://ftp.freetds.org/pub/freetds/stable/freetds-#{version}.tar.gz"
configure_options = "--with-tdsver=#{node['freetds']['tds_version']} #{'--disable-odbc' unless node['freetds']['odbc']}"

remote_file "#{Chef::Config[:file_cache_path]}/freetds-#{version}.tar.gz" do
  action :create_if_missing
  backup false
  source freetds_url
  checksum node['freetds']['checksum'] if node['freetds']['checksum']
end

bash 'build freetds' do
  cwd Chef::Config[:file_cache_path]
  code <<-EOF
    tar zxf freetds-#{version}.tar.gz
    (cd freetds-#{version} && CFLAGS='-fPIC' ./configure #{configure_options})
    (cd freetds-#{version} && make && make install)
  EOF
  not_if 'which tsql'
end


recipes/package.rb
#
# Cookbook Name:: freetds
# Recipe:: package
#
# Copyright (C) 2014 Olivier Brisse
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

case node['platform_family']
when 'debian'
  include_recipe 'apt::default'

  node['freetds']['packages'].each do |pkg|
    package pkg
  end
else
  raise 'Unsupported plaftorm, use the source install method'
end


attributes/default.rb
#
# Cookbook Name:: freetds
# Attributes:: default
#
# Copyright (C) 2013 Olivier Brisse
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

default['freetds']['install_method'] = 'package'
default['freetds']['version']        = '0.95.87'
default['freetds']['checksum']       = '0762b8947442ca355d85ce8e74752067'
default['freetds']['packages']       = %w(freetds-bin freetds-dev)
default['freetds']['tds_version']    = '7.1'
default['freetds']['odbc']           = false
default['freetds']['text_size']      = 64_512
default['freetds']['client_charset'] = nil

case node['freetds']['install_method']
when 'package'
  default['freetds']['dir']          = '/etc/freetds'
when 'source'
  default['freetds']['dir']          = '/usr/local/etc'
end

default['freetds']['servers'] = [
  {
    'name' => 'egServer50',
    'description' => 'A typical Sybase server',
    'host' => 'symachine.domain.com',
    'port' => 5000,
    'tds_version' => '5.0',
    'encryption' => 'off'
  },
  {
    'name' => 'egServer70',
    'description' => 'A typical Microsoft server',
    'host' => 'ntmachine.domain.com',
    'port' => 1433,
    'tds_version' => '7.0',
    'client_charset' => 'UTF-8',
    'encryption' => 'off'
  }
]
